lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs

```diff
@@ -0,0 +1,285 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Provides the 'TokenBundle' type, which combines a 'Coin' (lovelace) value
+--   with a map of named token quantities, scoped by token policy.
+--
+-- This module is meant to be imported qualified. For example:
+--
+-- >>> import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TB
+--
+module Cardano.Wallet.Primitive.Types.TokenBundle
```

Comment:
@rvl wrote:
> rolling TokenQuantity and TokenMap into this module - TokenBundle
> If you believe splitting it up will make compile times faster, then prove it.

Compilation time is not the major motivation for having separate modules here.

`TokenMap` and `TokenBundle` are separate modules for the following reasons:

1. They fulfill different purposes:

    a. `TokenMap` defines a map of **_non-ada_** quantities, for situations where ada quantities are forbidden (such as the `mint` field of a transaction).

    b. `TokenBundle` adds an ada quantity and defines a set of extra operations for converting between bundles and `Coin` values.

2. `TokenMap` has an invariant. By encapsulating this in its own module and not exporting the public constructor:

    * we can limit the number of operations that potentially violate the invariant, requiring fewer tests.
    * we can avoid the complication of considering `Coin` values when writing tests for `TokenMap`, as it can't include them.

4. `TokenMap` and `TokenBundle` have some common operations (for example, `getQuantity` and `adjustQuantity`) but they operate on different types.

    Moving them into the same module means we'd need a way to distinguish between these functions. We could use names that include the type they operate on, such as  `adjustTokenMapQuantity` and `adjustTokenBundleQuantity`, but this seems awkward. By putting them in different modules, they can have separate namespaces, removing the clash.

@rvl wrote:
> so many little modules

I'm not sure I agree that `TokenBundle` or `TokenMap` are small modules:

| Module | Size |
| -- | -- |
| `TokenMap` | 521 lines |
| `TokenBundle` | 285 lines |

I agree that `TokenQuantity` is relatively small, but it's not smaller than `Coin`, which is also separate. Because `TokenQuantity` is used by both `TokenMap` and `TokenBundle`, there doesn't seem to be a good reason to choose one or the other, unless we allow re-exports, which many people dislike due to code navigation.

Reply:
Even with go-to-definition support in your editor (even github may support it soon too), having less files is easier to work with.

Yes TokenMap and TokenBundle are different things, but is TokenMap ever going to be used independently of TokenBundle?

It's just lots of lines of fluff.

Reply:
@rvl wrote:
> Yes TokenMap and TokenBundle are different things, but is TokenMap ever going to be used independently of TokenBundle?

**Yes.** I actually answered that in my [earlier comment](https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r542176792):

> `TokenMap` defines a map of non-ada quantities, for situations where ada quantities are forbidden (**such as the `mint` field of a transaction**).

Users have a choice between:
* `TokenBundle`
   allows ada quantities (used by `TxOut.tokens`).
* `TokenMap`
   does not allow ada quantities (will be used by `Tx.mint`, once we support minting).

@rvl wrote:
> Even with go-to-definition support in your editor (even github may support it soon too), having less files is easier to work with.

I'm sorry, I'm not really persuaded by this argument. If that were true, why not just put everything in one large file?

Of course, there has to be a balance somewhere.

The fact that `TokenMap` has an invariant is a good reason to encapsulate this type in its own module. Not exporting the default constructor means that callers are forced to use smart constructors, which we test in `TokenMapSpec`. 

A side benefit is that tests for `TokenBundle` don't have to care about preserving the invariant of `TokenMap`, because the responsibility is encapsulated in that module (and its tests).

I could write more but I'd just be repeating my [earlier comment](https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r542176792).

> It's just lots of lines of fluff.

Let's keep this review productive, shall we? :)

Reply:
If you want to dig in and throw around "moving forward", then there's not much I can do, other than say that I'm trying to help you keep things simple.

Perhaps the data type for mint/burn ought to be different to the one for UTxO value? One of these is non-negative. You could consider removing the TokenQuantity module (fluff) and parameterise TokenMap by Integer or Natural. Up to you.

Reply:
> If you want to dig in and throw around "moving forward", then there's not much I can do, other than say that I'm trying to help you keep things simple.

First of all, I appreciate that you're trying to help. Reviewing a large PR is not easy.

But surely you must realize that comments like "It's just lots of lines of fluff" are not helpful. This comes across as a bit condescending, especially when someone has put thought and effort into a design. If you have doubts about a particular design choice, why not just reach out to me and ask? I'm happy to answer questions about anything. I'm also happy to change things (within reason) if there's a clear advantage to doing things another way.

I'd really like to overcome this hurdle and work with you on this.

When you asked me to justify the separation of `TokenMap` and `TokenBundle` (which is mentioned in the design), I gave several reasons, in good faith:

1. Encapsulating the invariant of `TokenMap`.
2. Tests for `TokenMap` can focus on properties of the map (including the invariant), without the additional complication of ada `Coin` values.
3. Separation of concerns and responsibilities. `TokenMap` is responsible for the map structure, and `TokenBundle` is responsible for interoperability with `Coin` values.
4. Allowing the functions of each type to have their own separate namespaces (if you want an analogy, think of `Map` and `NonEmptyMap`, which share similar sets of operations, but operate on different types.).

But as I can tell, you've not answered these points. It's exhausting to participate in a review where the reviewer doesn't respond to points raised.

This whole thread started with:
> Please would you consider following my suggestion from yesterday and rolling TokenQuantity and TokenMap into this module - TokenBundle. Because we don't need so many little modules.

I get that you have a preference for larger modules, which I respect. But `TokenMap` and `TokenBundle` are not small modules.

After some to and fro, you mentioned that:
> having less files is easier to work with.

This seems very subjective. Where do you draw the line?

I think the following are far more important than module size:

- encapsulation of invariants.
- prevention of regressions.
- separation of concerns.
- clear name-spacing.

Should we sacrifice any of these just for the sake of dropping our module count by one?

> Perhaps the data type for mint/burn ought to be different to the one for UTxO value? One of these is non-negative. You could consider removing the TokenQuantity module (fluff) and parameterise TokenMap by Integer or Natural. Up to you.

This is actually a good point. Though I don't understand the need to use antagonistic language like "fluff". Why is that necessary? As for choice to put `TokenQuantity` in a `newtype`: this was in the design document.

**As for parameterization**: I agree we should at least consider this. I propose that the right time to discuss this will come when we design support for minting. For now, we're probably not in danger of creating negative quantities, as nothing in the wallet creates real token bundles (yet). Perhaps we can settle on `Integer` or `Natural`, and then parameterize later, if we need to?

Reply:
It's not worth my time to rebut every dubious point, if I can see that you feel very strongly about this minor issue. Your PR - do as you wish.
For all this talk of invariants, the only invariant for TokenMap is that values are non-zero - but what happens if that is ever violated? Nothing probably! The wallet will very likely continue to function as normal. That invariant may even make implementing coin selection more difficult. And yet there is no invariant preventing negative token values!? This is a priority inversion due to overfluffing the fluff with too much fluff. Fluff is a technical term meaning low function to LOC ratio.

Reply:
> the only invariant for TokenMap is that values are non-zero - but what happens if that is ever violated?

The `Eq` instance will become incorrect, and it might even be the case that the minimum ada calculation (provided by the ledger) will give incorrect results (though I would need to double-check this).

Of course, we could rewrite our `Eq` instance to handle zero values, but that seems pointless given we have an auto-derived one that already works.

> The wallet will very likely continue to function as normal.

The wallet will also function as normal if the `TokenBundle` and `TokenMap` code is organized in 2 modules rather than 1, right? Merging modules (which was the reason for this thread) creates extra work here.

> That invariant may even make implementing coin selection more difficult. 

Can you say why the invariant would make coin selection more difficult?

The `getQuantity` function returns `0` in the case where there is no quantity for a given token, so callers shouldn't need to do anything special in the case where a bundle doesn't have a particular token.

> And yet there is no invariant preventing negative token values!?

AFAIK, this doesn't affect the integrity of the map, or the `Eq` instance.

Both [`cardano-ledger-specs`](https://github.com/input-output-hk/cardano-ledger-specs/blob/6e30ecf73c9296bc320d71f3a6b7bfad3ce117e9/shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs#L104) and [`cardano-node`](https://github.com/input-output-hk/cardano-node/blob/e20559ba6288d09bdc22a44cba7806ff85111274/cardano-api/src/Cardano/Api/Value.hs#L105) also use `Integer` here, presumably so that these types can be reused in the context of minting and burning.

But as you pointed out, we can parameterize this in a later PR, so that:
- we can use `Natural` in the case of a `TokenBundle` that's included in a transaction output.
- we can use `Integer` in the case of a non-ada `TokenMap` included in the `mint` field of a transaction.